numpy/lib/npyio.py

```diff
@@ -784,6 +783,162 @@ def floatconv(x):
+                      f'it contains at least one element of type '
+                      f'{type(c)}',)
+            raise
+
+    def skipper(n):
+        normed_skipcols = []
+        for c in skipcols:
+            try:
+                newc = normalize_axis_index(c, n)
+            except np.AxisError:
+                raise IndexError(f'skip column index {c} out of range for '
+                                 f'file with {n} columns') from None
+            normed_skipcols.append(newc)
+
+        usecols = [c for c in range(n) if c not in normed_skipcols]
+        return usecols
+
+    return skipper
+
+
+def _normalize_usecols_to_list(usecols, n):
+    """
+    Convert the argument passed in `usecols` to a list of column indices.
+
+    This function is used in `loadtxt` and `genfromtxt` to process the
+    `usecols` argument.  The function converts `usecols` to a list of int.
+
+    Parameters
+    ----------
+    usecols : int, sequence of int, slice or callable
+        If `usecols` is a callable, it must have signature::
+
+            f(n: int) -> List[int]
+
+    n : int
+        The actual number of fields found in the file.
+
+    Examples
+    --------
+    >>> _normalize_usecols_to_list([1, 2, -1], 10)
+    [1, 2, -1]
+    >>> _normalize_usecols_to_list(3, 10)
+    [3]
+    >>> _normalize_usecols_to_list(slice(1, 10, 2), 10)
+    [1, 3, 5, 7, 9]
+    >>> normalize_usecols_to_list(lambda n: list(range(n // 2)), 10)
+    [0, 1, 2, 3, 4]
+    """
+    if isinstance(usecols, slice):
+        return list(range(n)[usecols])
+
+    func = False
+    if callable(usecols):
+        usecols = usecols(n)
+        func = True
+
+    try:
+        usecols = list(usecols)
+    except TypeError:
+        # Attempt to convert to list failed, so the only remaining option
+        # is that usecols is an integer.
+        usecols = [usecols]
+
+    result = []
+    for c in usecols:
+        try:
+            c = opindex(c)
+        except TypeError as e:
+            if func:
+                e.args = (f"usecols must be an int or a sequence of ints but "
+                          f"the function provided returned a value that "
+                          f"contains at least one element of type "
+                          f"{type(c)}",)
+            else:
+                e.args = (f"usecols must be an int or a sequence of ints but "
+                          f"it contains at least one element of type "
+                          f"{type(c)}",)
+            raise
+        try:
+            newc = normalize_axis_index(c, n)
+        except np.AxisError:
+            msg = f'column index {c} out of range for file with {n} columns'
+            if func:
+                func_premessage = ("error in the values returned by the "
+                                   "callable passed to 'usecols': ")
+                msg = func_premessage + msg
+            raise IndexError(msg) from None
```

Comment:
This sort of feels like an abuse of `normalize_axis_index`, which was intended only for axes.

Notably, axes are limited to `int`, but indices can be larger than that.

Reply:
`normalize_axis_index` maps to `check_and_adjust_axis_msg`.

`check_and_adjust_index` is the internal function that you should be using here, but that's not expose to python

Reply:
Yup, a convenient above.  It will be easy enough to add a local private function, say `_normalize_index(k, n)` that does basically the same thing, but raises `IndexError` instead of `AxisError`.  Options for customizing the error message could be added so there would be no need to catch and reraise the exception.

Reply:
Not a valid abuse though, the `int` / `intp` mismatch is a dealbreaker. `range(n)[c]` is perhaps the easiest way to spell what you want.

Reply:
OK, I'm dropping the use of `normalize_axis_index`.  Not having to convert the `AxisError` to something else will clean up the code a bit.